File: app/src/main/java/kittentrate/data/repository/GameRepository.java
package kittentrate.data.repository;

import java.util.List;

import kittentrate.data.rest.NetworkCallback;
import kittentrate.score.PlayerScore;

/**
 * Created by Manuel Lorenzo on 18/03/2017.
 */
public class GameRepository implements GameDataSource.LocalDataSource, GameDataSource.SharedPreferencesDataSource, GameDataSource.Repository {
    private GameDataSource.RemoteDataSource remoteDataSource;
    private GameDataSource.LocalDataSource localDataSource;
    private GameDataSource.SharedPreferencesDataSource preferencesManager;
    private static GameRepository gameRepository;

    public static GameRepository getInstance(GameDataSource.LocalDataSource gameLocalDataSource, GameDataSource.RemoteDataSource gameRemoteDataSource, GameDataSource.SharedPreferencesDataSource manager) {
        if (gameRepository == null) {
            gameRepository = new GameRepository(gameLocalDataSource, gameRemoteDataSource, manager);
        }
        return gameRepository;
    }

    private GameRepository(GameDataSource.LocalDataSource gameLocalDataSource, GameDataSource.RemoteDataSource gameRemoteDataSource, GameDataSource.SharedPreferencesDataSource manager) {
        remoteDataSource = gameRemoteDataSource;
        localDataSource = gameLocalDataSource;
        preferencesManager = manager;
    }

    @Override
    public List<PlayerScore> getTopScores() {
        return localDataSource.getTopScores();
    }

    @Override
    public long addTopScore(PlayerScore playerScore) {
        return localDataSource.addTopScore(playerScore);
    }

    @Override
    public void deleteAllScores() {
        localDataSource.deleteAllScores();
    }

    @Override
    public void setPreferencesPhotoTag(String photoTag) {
        preferencesManager.setPreferencesPhotoTag(photoTag);
    }

    @Override
    public String getPreferencesPhotoTag() {
        return preferencesManager.getPreferencesPhotoTag();
    }

    @Override
    public void getPhotos(NetworkCallback networkCallback) {
        String photoTag = preferencesManager.getPreferencesPhotoTag();
        remoteDataSource.getPhotos(photoTag, networkCallback);
    }
}

File: app/src/main/java/kittentrate/data/repository/GameRepository.kt
package kittentrate.data.repository

import io.reactivex.Observable
import kittentrate.GameApplication
import kittentrate.data.mapping.PhotoEntityMapper
import kittentrate.data.repository.model.FlickrPhoto
import kittentrate.data.repository.model.PhotoEntity
import kittentrate.score.PlayerScore
import kittentrate.utils.applySchedulers

/**
 * Created by Manuel Lorenzo on 18/03/2017.
 */
class GameRepository private constructor(private val localDataSource: GameDataSource.LocalDataSource,
                                         private val remoteDataSource: GameDataSource.RemoteDataSource,
                                         private val preferencesManager: GameDataSource.SharedPreferencesDataSource)
    : GameDataSource.LocalDataSource, GameDataSource.SharedPreferencesDataSource, GameDataSource.Repository {

    override fun getTopScores(): Observable<List<PlayerScore>> {
        return Observable
                .fromCallable({
                    GameApplication.database
                            .playerScoresDao().getTopScores()
                })
                .applySchedulers()
    }

    override fun addTopScore(playerScore: PlayerScore): Long =
            GameApplication.database.playerScoresDao()
                    .addTopScore(playerScore)

    override fun deleteAllScores() {
        localDataSource.deleteAllScores()
    }

    override fun setPreferencesPhotoTag(photoTag: String) {
        preferencesManager.preferencesPhotoTag = photoTag
    }

    override fun getPreferencesPhotoTag(): String = preferencesManager.preferencesPhotoTag

    override fun getPhotos(): Observable<List<PhotoEntity>> {
        val serviceMapper = PhotoEntityMapper()
        val photoTag = preferencesManager.preferencesPhotoTag
        return remoteDataSource.
                getPhotos(photoTag)
                .flatMap({ t: FlickrPhoto -> Observable.just(serviceMapper.transform(t.photos.photo)) })
                .applySchedulers()
    }

    companion object {
        private var gameRepository: GameRepository? = null

        fun getInstance(gameLocalDataSource: GameDataSource.LocalDataSource,
                        gameRemoteDataSource: GameDataSource.RemoteDataSource,
                        sharedPreferencesDataSource: GameDataSource.SharedPreferencesDataSource): GameRepository {
            if (gameRepository == null) {
                gameRepository = GameRepository(gameLocalDataSource, gameRemoteDataSource, sharedPreferencesDataSource)
            }
            return gameRepository as GameRepository
        }

        fun destroy() {
            gameRepository = null
        }
    }
}
